### PR TITLE
internal/exp/metrics: identity types

### PR DESCRIPTION
**Pull request by tombrk**
Adds a new internal, _experimental_ package `metrics/identity` which implements identity types for resource, scope, metric and stream.

This is closely related to work being done in #30707 and #30827.

The package is specifically experimental, as it shall be treated as an internal component to above processors which may change at any moment as long as those are under active initial development.

/cc @jpkrohling @djaglowski @RichieSams

### CODE REVIEW

File: internal/exp/metrics/identity/resource.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"hash"
+	"hash/fnv"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+type Resource struct {
+	attrs [16]byte
```
Comment (RichieSams) [inline internal/exp/metrics/identity/resource.go:15]: Do we need to include the resource SchemaURL as part of the identifier?

File: internal/exp/metrics/identity/scope.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"hash"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+type Scope struct {
+	res Resource
+
+	name    string
```
Comment (RichieSams) [inline internal/exp/metrics/identity/scope.go:16]: Do we need to include the scope schemaURL as part of the identifier?

**Comment by RichieSams** [comment]
Added two small comments. Overall, I like the approach a lot; the code design is very clean

File: internal/exp/metrics/identity/metric.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	sum.Write([]byte(i.unit))
+
+	var mono byte
+	if i.monotonic {
+		mono = 1
+	}
+	sum.Write([]byte{byte(i.ty), mono, byte(i.temporality)})
+	return sum
+}
+
```
Comment (RichieSams) [inline internal/exp/metrics/identity/metric.go:36]: Can we add functions for each identity struct to return the "inner" ID? For example:

```
func (i Metric) Scope() Scope {
    return i.scope
}
```

So `Stream` has a `Metric()`, `Metric` has a `Scope()`, and `Scope` has a `Resource()`.

**Comment by RichieSams** [comment]
@jpkrohling @djaglowski Any thoughts on this PR?

@sh0rez What do you think of my comments about the schemaURL and potentially allowing exposing the "inner" ID?

File: internal/exp/metrics/identity/resource.go (hunk 1 above)
Comment (sh0rez) [reply internal/exp/metrics/identity/resource.go:15]: no idea, can someone clarify? @jpkrohling @djaglowski

Comment (jpkrohling) [reply internal/exp/metrics/identity/resource.go:15]: Yes: https://opentelemetry.io/docs/specs/otel/resource/sdk/#create

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -100,7 +100,7 @@ extension/storage/filestorage/                           @open-telemetry/collect
 
 internal/aws/                                            @open-telemetry/collector-contrib-approvers @Aneurysm9 @mxiamxia
 internal/docker/                                         @open-telemetry/collector-contrib-approvers @rmfitzpatrick @jamesmoessis
-
+internal/exp/metrics/                                    @open-telemetry/collector-contrib-approvers @sh0rez
```
Comment (jpkrohling) [inline .github/CODEOWNERS:103]: @RichieSams, how do you feel about being a code owner of this with @sh0rez ?

**Review by jpkrohling — COMMENTED** [review]
I'm OK with this change, but it should be discussed with other folks owning metrics components. I also feel like this should be part of core, not contrib.

cc @dmitryax, @codeboten, @mx-psi

**Comment by djaglowski** [comment]
> I'm OK with this change, but it should be discussed with other folks owning metrics components. 

This looks pretty good to me. Having a dedicated package which handles uniqueness according to our data model will be very helpful. I've seen numerous other implementations, most of which tend to be suboptimal. As a shared package, it can evolve towards correctness and performance without otherwise being a concern of those who work on it. 

> I also feel like this should be part of core, not contrib.

I agree but in my opinion it makes sense to develop it in contrib and move it to core when it is proven out.

File: internal/exp/metrics/identity/resource.go (hunk 1 above)
Comment (djaglowski) [reply internal/exp/metrics/identity/resource.go:15]: This came up in the maintainers meeting yesterday and consensus seemed to be that most language libraries are not setting this field. Technically it is optional so I would be ok with leaving it out for now and adding later if necessary.

File: internal/exp/metrics/identity/scope.go (hunk 2 above)
Comment (djaglowski) [reply internal/exp/metrics/identity/scope.go:16]: Leave it out IMO

File: .github/CODEOWNERS (hunk 4 above)
Comment (RichieSams) [reply .github/CODEOWNERS:103]: That's fine by me. I already was adding myself in the followup PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31089

Comment (djaglowski) [reply .github/CODEOWNERS:103]: @RichieSams, you'll have to become a member of the OpenTelemetry org in order to be a code owner. [Requirements are here](https://github.com/open-telemetry/community/blob/main/community-membership.md#requirements). I'm happy to be one of your sponsors.

**Review by mx-psi — COMMENTED** [review]
Makes sense to me to have this. Would be good to closely follow the spec, since there are some subtleties involved, and it's also a good opportunity to clarify the spec if it's unclear that a particular attribute is part of the metrics' identity.

File: .github/CODEOWNERS (hunk 4 above)
Comment (sh0rez) [reply .github/CODEOWNERS:103]: that would apply to me as well then. @djaglowski would you be willing to be a sponsor for me as well?

Comment (RichieSams) [reply .github/CODEOWNERS]: @jpkrohling Would you be willing to be our second sponsor?

Comment (jpkrohling) [reply .github/CODEOWNERS]: Absolutely

Comment (RichieSams) [reply .github/CODEOWNERS]: https://github.com/open-telemetry/community/issues/1945

**Comment by RichieSams** [comment]
@sh0rez It looks like some CI checks are failing:
```
Porto links are out of date, please run "make goporto" and commit the changes in this PR
```

@djaglowski @jpkrohling @mx-psi Once that is fixed, do you have any other comments? Is it ok to merge?

File: internal/exp/metrics/metadata.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+status:
+  codeowners:
+    active: [sh0rez]
```
Comment (jpkrohling) [inline internal/exp/metrics/metadata.yaml:3]: Isn't this the source of the codeowners file?

Comment (sh0rez) [reply internal/exp/metrics/metadata.yaml]: the autogen does not work for me (not an org member, so the script cannot fetch the member list). according to contributing.md you have do edit manually then. 

updated the yaml to reflect the codeowners file

**Comment by sh0rez** [comment]
@RichieSams as a future codeowner, could you approve this PR as well? :)

**Review by RichieSams — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**